File: common-system-test/src/main/java/com/teamscale/test/commons/Session.java
package com.teamscale.test.commons;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.teamscale.client.EReportFormat;
import com.teamscale.client.JsonUtils;
import com.teamscale.report.compact.TeamscaleCompactCoverageReport;
import com.teamscale.report.testwise.model.TestwiseCoverageReport;
import org.conqat.lib.commons.collections.ListMap;
import spark.Request;

import java.io.IOException;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Represents an upload session to Teamscale which contains data from a single partition, for a specific
 * commit/revision. A session may contain data in multiple formats.
 */
public class Session {

	private final String partition;
	private final String revision;
	private final String commit;
	private final String repository;
	private boolean committed;

	private final ListMap<EReportFormat, ExternalReport> reports = new ListMap<>();

	public Session(Request request) {
		partition = request.queryParams("partition");
		revision = request.queryParams("revision");
		commit = request.queryParams("t");
		repository = request.queryParams("repository");
	}

	/**
	 * Retrieves the commit information for the session. Combines the revision and commit data into a single string
	 * representation separated by a comma.
	 */
	public String getCommit() {
		return revision + ":" + repository + ", " + commit;
	}

	/** Returns the partition name for which the session was opened. */
	public String getPartition() {
		return partition;
	}

	/**
	 * Marks the session as committed, which means no more data will be added, and a real Teamscale would start
	 * processing the data now.
	 */
	public void markCommitted() {
		committed = true;
	}

	/** Whether the session was commited. */
	public boolean isCommitted() {
		return committed;
	}

	/** Adds a new report into the session. */
	public void addReport(String format, String report) {
		reports.add(EReportFormat.valueOf(format), new ExternalReport(report));
	}

	/** Returns all reports form this session. */
	public List<ExternalReport> getReports() {
		return reports.getValues();
	}

	/** Returns all reports of the given format. */
	public List<ExternalReport> getReports(EReportFormat format) {
		return reports.getCollection(format);
	}

	/** Returns the only report in the given format. It asserts that there are no other reports present. */
	public String getOnlyReport(EReportFormat format) {
		if (reports.getKeys().size() != 1) {
			throw new AssertionError("Expected exactly one report format, but got " + reports.getKeys() + "!");
		}
		if (!reports.containsCollection(format)) {
			throw new AssertionError(
					"No " + format.getReadableName() + " report found! Session contains " + reports.getKeys()
							.stream().map(EReportFormat::getReadableName)
							.collect(Collectors.toSet()) + " reports.");
		}
		if (reports.getCollection(format).size() != 1) {
			throw new AssertionError(
					"Expected exactly one " + format.getReadableName() + " report, but got " +
							reports.getCollection(format).size() + ".");
		}
		return reports.getCollection(format).get(0).getReportString();
	}

	/**
	 * Returns the report at the given index in {@link #reports}, parsed as a {@link TestwiseCoverageReport}.
	 *
	 * @throws IOException when parsing the report fails.
	 */
	public TestwiseCoverageReport getTestwiseCoverageReport(int index) throws IOException {
		return JsonUtils.deserialize(
				reports.getCollection(EReportFormat.TESTWISE_COVERAGE).get(index).getReportString(),
				TestwiseCoverageReport.class);
	}

	/** Returns the only Testwise Coverage report in deserialized form. */
	public TestwiseCoverageReport getOnlyTestwiseCoverageReport() throws JsonProcessingException {
		return JsonUtils.deserialize(
				getOnlyReport(EReportFormat.TESTWISE_COVERAGE),
				TestwiseCoverageReport.class);
	}

	/** Returns the only Compact Coverage report in deserialized form. */
	public TeamscaleCompactCoverageReport getOnlyCompactCoverageReport() throws JsonProcessingException {
		return JsonUtils.deserialize(
				getOnlyReport(EReportFormat.TEAMSCALE_COMPACT_COVERAGE),
				TeamscaleCompactCoverageReport.class);
	}

	/**
	 * Returns the report at the given index in {@link #reports}, parsed as a {@link TeamscaleCompactCoverageReport}.
	 *
	 * @throws IOException when parsing the report fails.
	 */
	public TeamscaleCompactCoverageReport getCompactCoverageReport(int index) throws IOException {
		return JsonUtils.deserialize(
				reports.getCollection(EReportFormat.TEAMSCALE_COMPACT_COVERAGE).get(index).getReportString(),
				TeamscaleCompactCoverageReport.class);
	}
}

File: common-system-test/src/main/kotlin/com/teamscale/test/commons/Session.kt
package com.teamscale.test.commons

import com.fasterxml.jackson.core.JsonProcessingException
import com.teamscale.client.EReportFormat
import com.teamscale.client.JsonUtils
import com.teamscale.report.compact.TeamscaleCompactCoverageReport
import com.teamscale.report.testwise.model.TestwiseCoverageReport
import org.conqat.lib.commons.collections.ListMap
import spark.Request
import java.io.IOException
import java.util.stream.Collectors

/**
 * Represents an upload session to Teamscale, which contains data from a single partition, for a specific
 * commit/revision. A session may contain data in multiple formats.
 */
class Session(request: Request) {
	/** Returns the partition name for which the session was opened.  */
	val partition: String? = request.queryParams("partition")
	private val revision: String? = request.queryParams("revision")
	private val commit: String? = request.queryParams("t")
	private val repository: String? = request.queryParams("repository")

	/** Whether the session was committed.  */
	var isCommitted: Boolean = false
		private set

	private val reports = ListMap<EReportFormat, ExternalReport>()

	/**
	 * Retrieves the commit information for the session. Combines the revision and commit data into a single string
	 * representation separated by a comma.
	 */
	fun getCommit() = "$revision:$repository, $commit"

	/**
	 * Marks the session as committed, which means no more data will be added, and a real Teamscale would start
	 * processing the data now.
	 */
	fun markCommitted() {
		isCommitted = true
	}

	/** Adds a new report into the session.  */
	fun addReport(format: String, report: String) {
		reports.add(EReportFormat.valueOf(format), ExternalReport(report))
	}

	/** Returns all reports from this session.  */
	fun getReports(): List<ExternalReport> = reports.getValues()

	/** Returns all reports of the given format.  */
	fun getReports(format: EReportFormat): MutableList<ExternalReport> = reports.getCollection(format)

	/** Returns the only report in the given format. It asserts that there are no other reports present.  */
	fun getOnlyReport(format: EReportFormat): String {
		check(reports.getKeys().size == 1) {
			"Expected exactly one report format, but got ${reports.getKeys()}!"
		}
		check(reports.containsCollection(format)) {
			"No ${format.readableName} report found! Session contains ${
				reports.getKeys().map(EReportFormat::readableName).toSet()
			} reports."
		}
		check(reports.getCollection(format).size == 1) {
			"Expected exactly one ${format.readableName} report, but got ${reports.getCollection(format).size}."
		}
		return reports.getCollection(format).first().reportString
	}

	/**
	 * Returns the report at the given index in [reports], parsed as a [TestwiseCoverageReport].
	 *
	 * @throws IOException when parsing the report fails.
	 */
	@Throws(IOException::class)
	fun getTestwiseCoverageReport(index: Int) =
		reports.getCollection(EReportFormat.TESTWISE_COVERAGE).getOrNull(index)?.let {
			JsonUtils.deserialize<TestwiseCoverageReport>(it.reportString)
		}

	@get:Throws(JsonProcessingException::class)
	val onlyTestwiseCoverageReport: TestwiseCoverageReport
		/** Returns the only Testwise Coverage report in deserialized form.  */
		get() = JsonUtils.deserialize<TestwiseCoverageReport>(
			getOnlyReport(EReportFormat.TESTWISE_COVERAGE)
		)

	@get:Throws(JsonProcessingException::class)
	val onlyCompactCoverageReport: TeamscaleCompactCoverageReport
		/** Returns the only Compact Coverage report in deserialized form.  */
		get() = JsonUtils.deserialize<TeamscaleCompactCoverageReport>(
			getOnlyReport(EReportFormat.TEAMSCALE_COMPACT_COVERAGE)
		)

	/**
	 * Returns the report at the given index in [reports], parsed as a [TeamscaleCompactCoverageReport].
	 *
	 * @throws IOException when parsing the report fails.
	 */
	@Throws(IOException::class)
	fun getCompactCoverageReport(index: Int) =
		reports.getCollection(EReportFormat.TEAMSCALE_COMPACT_COVERAGE).getOrNull(index)?.let {
			JsonUtils.deserialize<TeamscaleCompactCoverageReport>(it.reportString)
		}
}
